### DISCUSSION TITLE
Expected behaviour when indenting twice (two tabs) in admonitions?

### DISCUSSION THREAD

**Discussion by StarfallProjects** [squidfunk/mkdocs-material#6168] [Q&A] [answered]
A visual bug recently appeared: admonition content being displayed as code (example: https://docs.n8n.io/user-management/ldap/). I can't pinpoint exactly when it started, but only the last few days.

However, I am not sure if this is a bug with the theme, or intended behaviour, so before I file a bug report: is it intended that when you put two tabs in an admonition it renders as code?

For example, this inserts code formatting:

!!! note "an example note"
`<tab><tab>`Contents

While this does not:

!!! note "an example note"
`<tab>`Contents

I would have expected the tabs to be replaced with spaces (we actually had to do a workaround to prevent this in our code blocks), so am a bit surprised they're affecting it at all. The equivalent spaces (4) don't result in code formatting. I have done a minimal reproduction, and it isn't our workaround causing this - it occurs without it in place.

@squidfunk this is an n8n problem, not a personal site problem, in case that affects response time :-)

**Comment by StarfallProjects** [comment]
Ok the more I look at this the more I think not-bug. Python Markdown will be replacing two tabs with eight spaces?
You can probably ignore this - although the fact it only started being an issue recently still gives me pause.

**Reply by alexvoss** [reply]
Thanks for reaching out. I think this is a side effect of you being able to put code into an admonition. Indentation really does matter in Markdown.

(Speaking as someone who is not a fan of languages where indentation matters - did we run out of delimiters or were they just too difficult to reach on non-US/UK keyboards?)

**Comment by squidfunk** [comment]
Hmm, I'm pretty sure this has nothing to do with Material for MkDocs, but with Python Markdown extensions in general. Could you please try the `mkdocs` and `readthedocs` theme if they exhibit the same behavior? If yes, please report this to Python Markdown.

**Comment by facelessuser** [comment, marked as answer]
So, I will give some insight into the recent surfacing of this and state it is not a bug. Admonitions, in this format, have always required that the content be indented. Additionally, indented blocks are generally treated as code blocks in Markdown. So to have indented code blocks under an admonition you would have to indent twice.

Now, there was a bug that I recently fixed in admonitions. Indented code blocks were not treated properly. https://github.com/Python-Markdown/markdown/issues/1329

Now that this has been fixed, indented code blocks now work properly when they are the first block under admonitions, which is likely what you are seeing.

With all of that said, I've never been a fan of having to indent the _all_ content under Admonitions, Details, Definitions, Tabs, etc. Pymdown Extensions implemented plugins for details and tabs using this Admonition style approach because it was a bit complicated to create such blocks with complex levels of blocks under them under Python Markdown. Recently though, I have created a way to do so without requiring all of this indentation.

If you do not like these "indentation" style containers, you can use Pymdownx block plugins for Admonition, Details, Tabs, etc. https://facelessuser.github.io/pymdown-extensions/extensions/blocks/plugins/admonition/.

**Reply by StarfallProjects** [reply]
Thanks! That solves the mystery of the recent change.

**Reply by alexvoss** [reply]
Oh, me loves the blocks with delimiters! For admonitions, add `pymdownx.blocks.admonition` to the `markdown_extensions:` in addition to the things that should already be there and it works a treat in Material.

**Reply by StarfallProjects** [reply]
Yeah I just made a ticket to try these when I get a chance. Definitely less room for error!

**Reply by alexvoss** [reply]
I am working on an example, will share. Custom admonition will need an `extra.css` but the basics are there as the `div` for a custom admonition has the appropriate `custom` class.

**Reply by facelessuser** [reply]
The "block" style replacements for various plugins are 100% compatible with the old "indented" style. They should be compatible with existing styles.

**Reply by facelessuser** [reply]
What I mean is the output is the same even if the input syntax is different.

**Reply by alexvoss** [reply]
Yep, and it works nicely. Will make a PR for the Material for MkDocs examples but need to figure out how to flag this is based on a feature that is still experimental. Hope to get this done today, though, and post the link. Otherwise I might just make it a 'show and tell'.

**Reply by facelessuser** [reply]
The experimental label was mainly expecting a number of opinions coming in about the syntax, but I am yet to hear anything negative about it. At some point, I'm going to remove the experimental status.

I have no idea how many people are actually using it compared to the existing indented approach. People often use what they are used to. I imagine there may be people just not aware of it either.

If you add examples, I would definitely note that people should not enable a legacy plugin with the new block plugin at the same time. Because they both have the same output, both plugins can sometimes step on each other's toes trying to process the content. If you have old style admonitions and new style block admonitions enabled, the new admonitions will see the new block syntax, but the old style will try to process the content under the block. It's just how the extensions work.

**Comment by StarfallProjects** [comment]
@alexvoss did the PR to support the new admonition syntax go live yet? Had a quick scan through recent release notes and can't spot what version it's in. I'd love to get us upgraded to this as we keep hitting blips in our admonition formatting.

**Reply by alexvoss** [reply]
Hi, sorry, I will try to get this done now. It is not so much "support" but just and example how to use the block syntax that can already be used. Give me a couple of hours to create an example that cover the different uses - admonitions, tabs, etc.

**Reply by StarfallProjects** [reply]
Ah k, I was trying to use it and it's just rendering the raw markup, which made me think it wasn't supported yet. I shall check more closely for user error.

**Reply by StarfallProjects** [reply]
For example, this:

```
/// admonition | Weeks without large solar flares 
    type: note 

In this tutorial, you are working with live date. If you find there are no X class solar flares when you run the workflow, try replacing **X** in **Value 2** with either **A**, **B**, **C**, or **M**. 
///
```

Is producing this:
![image](https://github.com/squidfunk/mkdocs-material/assets/10284570/2004cf38-1212-415a-90f3-4d8b7606c01d)

**Reply by alexvoss** [reply]
You will need to remove the `pymdownx.admonition` line from your `markdown_extensions` in the `mkdocs.yml` and replace it with `pymdownx.blocks.admonition`. Hope that fixes it.

**Reply by StarfallProjects** [reply]
I currently have this, as in the Material docs:
```
  - admonition
  - pymdownx.details
  - pymdownx.superfences
   ```
 
 So to confirm, I should replace `admonition` with `pymdownx.blocks.admonition`?
 
 (I can also just wait for your docs & examples updates to be fully ready though - I'm sure this will be answered in them)

**Reply by facelessuser** [reply]
@StarfallProjects I currently use Material with the new blocks formats. This is what you need: https://github.com/facelessuser/pymdown-extensions/blob/main/mkdocs.yml#L175

Note that the new formats replace the old formats, so yes, replace `admonition` with the new setup. In my example, I register all the Material special types, though you can omit `settings` and `new` as those are custom.

Then you can just use:

```
/// note | title
content
///
```

If you do **not** register any types, then you must specify them when needed.

```
/// admonition | title
    type: note

content
///
```

Hopefully, that is clear.

**Reply by StarfallProjects** [reply]
@facelessuser perfect thank you!

**Reply by StarfallProjects** [reply]
One follow-up: what's the syntax for collapsible blocks with this? (previously `???`)

**Reply by facelessuser** [reply]
All of this is documented here: https://facelessuser.github.io/pymdown-extensions/extensions/blocks/plugins/details/

But in short, the setup is like so: https://github.com/facelessuser/pymdown-extensions/blob/main/mkdocs.yml#L191

Details allow you to register types like admonitions, but since Admonition and Details use the same types, there would be a conflict, so I just manually specify detail types when needed as I use admonitions far more than details:

```
/// details | summary
    type: note

content
////
```

**Reply by alexvoss** [reply]
@facelessuser a quick question: I note that [the documentation says that admonitions like some admonition types are predefined](https://facelessuser.github.io/pymdown-extensions/extensions/blocks/plugins/admonition/#usage) but this does not seem to be the case?

> As a shortcut, Admonitions registers a number of admonition blocks that attach common classes: note, attention, caution, danger, error, tip, hint, warning.

I note that in the example you link to above, you defined them all. Did you decide not to do this in the end - perhaps because it might get in the way of people wanting to define their own custom blocks with these names?

**Reply by facelessuser** [reply]
> @facelessuser a quick question: I note that [the documentation says that admonitions like some admonition types are predefined](https://facelessuser.github.io/pymdown-extensions/extensions/blocks/plugins/admonition/#usage) but this does not seem to be the case?

They are defined, see the conversation here: https://github.com/facelessuser/pymdown-extensions/discussions/1973#discussioncomment-7275905. I should specify the the defaults in the docs though.

> I note that in the example you link to above, you defined them all. Did you decide not to do this in the end - perhaps because it might get in the way of people wanting to define their own custom blocks with these names?

You either accept the default or you define your own. But you are not obligated to use the predefined types even if registered. You can always use:

```
/// admonition | title
    type: my-type

content
///
```

**Reply by alexvoss** [reply]
I think my confusion comes from the fact the Material for MkDocs has different admonitions defined by default compared to what you have in the plugin. It has `bug`, for example, but not `attention`. For the example I am writing, it would seem best to advise people to just define the list of admonition types they want to use, which is what I have at the moment anyway. This is a bit of a difference compared to the traditional syntax but I think it is totally worth it.

**Reply by facelessuser** [reply]
Material is a user of pymdownx, pymdownx is not a Material library, we don't bias things for Material just because it is popular. The discussion I linked explains where the defaults come from, but not all people who use pymdownx are Material users.

**Reply by alexvoss** [reply]
Of course, I am just trying to figure out for myself what to write in the example, which is for Material. I am not suggesting what you should change anything. Apologies if I came across as trying to do that.

**Reply by facelessuser** [reply]
I would just state that the defaults do not accommodate all of Material's special types by default; therefore, if shorthand names are desired, those names must be registered and provide an example showing a configuration defining all Material types.

**Reply by facelessuser** [reply]
I think most people would want the short names for admonitions, so it may be a matter of just saying we need to define these because the defaults don't match Material's needs, and just offer the example as the recommended approach.
